hathor/indexes/memory_mempool_tips_index.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2021 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING, Iterable, Iterator, Optional, Set, cast
+
+from structlog import get_logger
+
+from hathor.indexes.mempool_tips_index import MempoolTipsIndex
+from hathor.transaction import BaseTransaction, Transaction
+from hathor.transaction.storage.traversal import BFSWalk
+from hathor.util import not_none
+
+if TYPE_CHECKING:  # pragma: no cover
+    from hathor.transaction.storage import TransactionStorage
+
+logger = get_logger()
+
+
+class MemoryMempoolTipsIndex(MempoolTipsIndex):
+    _index: Set[bytes]
+
+    def __init__(self):
+        self._index = set()
+        self.log = logger.new()
+
+    def _remove_many(self, remove_txs: Iterable[bytes]) -> None:
+        for tx in iter(remove_txs):
+            if tx in self._index:
+                self._index.remove(tx)
+
+    def update(self, tx: BaseTransaction) -> None:
+        # A new tx/block added might cause a tx in the tips to become voided. For instance,
+        # there might be a tx1 a double spending tx2, where tx1 is valid and tx2 voided. A new block
+        # confirming tx2 will make it valid while tx1 becomes voided, so it has to be removed
+        # from the tips.
+        assert tx.hash is not None
+        assert tx.storage is not None
+        to_remove: Set[bytes] = set()
+        to_remove_parents: Set[bytes] = set()
+        tx_storage = tx.storage
+        for tip_tx in self.iter(tx_storage):
```

Comment:
I was very confused the first time I read this line, as Python has a `iter` function itself and I didn't notice the `self`. So I thought you were trying to create an iterator over the storage, which didn't make any sense. Is this an internal function? Can we call it `_iter` or something else?

Reply:
This isn't an internal function. It's the method the index uses to expose an iterator. It would be possible to implement `__iter__` instead (which would allow doing `for tip_tx in self`), however it needs to receive a `tx_storage`, therefore it can't use `__iter__`.

Reply:
Ah, I got what you mean. I think now it's only used by other methods from the same class (from what I could find), but in the future it can be used by others.